Add missing dependencies and extra info to optimizer step docs

cameel opened the pull request:
~Depends on #15053.~ Merged.

The current optimizer docs are pretty incomplete when it comes to documenting prerequisites and recommended steps to run before/after. I need up to date info to verify that the new sequence (#15030) satisfies these requirements so I went over the docs and the docstrings of the steps in the code, compared them and amended the docs where necessary.

I also compiled a table listing them, which is what I will actually refer to when analyzing the sequence - doing it by reading the docs is too unwieldy.

cameel committed:
Optimizer docs: Document LiteralRematerialiser and details of FullInliner's heuristic

cameel committed:
cheatsheet: List optimizer step dependencies

cameel commented on docs/cheatsheet.rst:
```diff
@@ -168,3 +168,42 @@ Modifiers
+Optimizer Step Dependencies
+===========================
+
+=========================================== ===================================================== =============================================================== ===========================================
+Step                                        Hard prerequisites                                    Steps that should run before                                    Steps that should run after
+=========================================== ===================================================== =============================================================== ===========================================
+:ref:`a <ssa-transform>`                    Disambiguator, ForLoopInitRewriter                    ExpressionSplitter, CommonSubexpressionEliminator               UnusedAssignEliminator
+:ref:`C <conditional-simplifier>`           Disambiguator                                         SSATransform, DeadCodeEliminator
+:ref:`c <common-subexpression-eliminator>`  Disambiguator, ForLoopInitRewriter                    ExpressionSplitter, SSATransform                                UnusedPruner, UnusedAssignEliminator
```
Initially I wanted to just add the extra info to the table we already have in the optimizer docs, but our current CSS only allows very narrow tables. I even had to cut full names here to avoid the scrollbar hiding most of the info.

The best compromise I found was to create a new table in the cheatsheet. I don't like it very much but I like other options even less. I could maybe put it in the optimizer docs, but having two tables side by side seems weird. This is the kind of summarized information that actually fits the purpose of the cheatsheet, it's just that this is not the info users will be reaching for often, it will be mostly useful to us.

The nice thing is that here I can sort the steps alphabetically by their abbreviations, which is much more useful than the original order when decoding the sequences.

ekpyron replied:
Now you have the large comment exactly on the first line I wanted to comment on :-D...
Why the ``SSATransform`` here? There *could* be different implementations for the data flow analyzer, which might mean that pseudo-SSA form helps - but currently, if anything, it may get more slower due to larger numbers of variables, won't it?

EDIT: Ah, it's in the previous docs about it. And I see the point made there - not sure that makes much of a difference in practice and in almost all cases having *ever* run an SSA transform will do that (rather than having to be in pseudo-SSA form), but alright.

cameel replied:
Yeah, this table is not new information. It only summarizes info already available in step docs (in some cases amended with some extra info I pulled from the docstrings). This is what it says currently:

> This step is especially efficient if the ExpressionSplitter is run
> before. If the code is in pseudo-SSA form,
> the values of variables are available for a longer time and thus we
> have a higher chance of expressions to be replaceable.

So you're basically saying that there's some nuance missing in the current docs - because they do make it sound like you might want to use `SSATransform` here. In that case I'll try to adjust them based on what you said and then remove this from the table.

> Now you have the large comment exactly on the first line I wanted to comment on :-D...

You know that you can make multiple separate review comments for the same line, right?

cameel replied:
I added your remarks to CSE description. I still left SSA in the table, but added a footnote explaining that it does not have to run before the step every time.

cameel commented on docs/internals/optimizer.rst:
```diff
@@ -411,6 +411,8 @@ its visibility and it is impossible to reference variables defined in a differen
 The benefit of this stage is that function definitions can be looked up more easily
 and functions can be optimized in isolation without having to traverse the AST completely.
 
+Prerequisites: Disambiguator.
```
Note that not all steps list disambiguator as a prerequisite even though its own description above says that all of them depend on it.

BTW, not sure if listing prerequisites is all that useful since we have very few hard prerequisites and those are hard-coded anyway. Maybe it would be fine to remove this info. For now though we do list that info so I amended it for steps where it was available in the docstrings.

cameel commented on docs/internals/optimizer.rst:
```diff
@@ -1018,6 +1058,12 @@ a structural level:
+
+Prerequisites: Disambiguator.
+
+Important: Can only be used on EVM code.
```
Not sure about usefulness of this either now that we only have an EVM backend, but since other steps already list this kind of info (e.g. `Important: Introduces EVM opcodes and thus can only be used on EVM code for now` for ControlFlowSimplifier) I added it where it was available for now.

ekpyron replied:
The code only has that, because we haven't removed all of it when removing the EWASM parts :-).

cameel replied:
I mean, this still info makes some sense unless we know for sure that there will never be backends other than EVM. I suspect it's incomplete so it may be not very useful even then, but I don't know for sure so I didn't want to remove it without some confirmation.

cameel replied:
Remarks about EVM removed from the docs and also from the docstrings.

cameel commented on docs/internals/optimizer.rst:
```diff
@@ -985,6 +1018,9 @@ CommonSubexpressionEliminator, because SSA will make sure that the variables
 will not change and the CommonSubexpressionEliminator re-uses exactly the same
 variable if the value is known to be the same.
 
+Works best on code without literal arguments, which is the case when the code is in the
+:ref:`expression-split <expression-splitter>` form.
```
I based this addition on the docstrings, which says this: https://github.com/ethereum/solidity/blob/fb11553a650f6e07299354a867fc80e1c1ba3c0b/libyul/optimiser/EqualStoreEliminator.h#L35

Since SSATransform on its own does nothing about literal arguments, this basically means that that ExpressionSplitter is recommended before, right?

clonker replied:
it uses `isSimpleStore` of the `DataFlowAnalyzer`, which checks if the key and value arguments are identifiers, so indeed literals as arguments would fly under the radar and expression split form is recommended. This is also a place where destroying the strict expression split form would hurt (f ex in load resolver or literal rematerializer).

cameel commented on docs/cheatsheet.rst:
```diff
@@ -168,3 +168,42 @@ Modifiers
+:ref:`o <for-loop-init-rewriter>`           Disambiguator
+:ref:`p <unused-function-parameter-pruner>` Disambiguator, FunctionHoister                        LiteralRematerialiser                                           ExpressionInliner
+:ref:`r <unused-assign-eliminator>`         Disambiguator, ForLoopInitRewriter
+:ref:`S <unused-store-eliminator>`          Disambiguator, ForLoopInitRewriter                    SSATransform
```
Based on the discussion we had with @nikola-matic in https://github.com/ethereum/solidity/pull/14779#discussion_r1457976047, I think that CommonSubexpressionEliminator should be recommended before UnusedStoreEliminator. Anyone can confirm/deny?

cameel commented on docs/internals/optimizer.rst:
```diff
@@ -1117,9 +1168,9 @@ and replace all references to ``f`` by ``f2``.
 The inliner should be run afterwards to make sure that all references to ``f2`` are replaced by
 ``f``.
```
I'm assuming this means `ExpressionInliner`, because it inlines functions of exactly this form that this step introduces. Is that correct?

ekpyron replied:
No, that's just coincidence here - if you have a function with three return arguments and only one of them is unused, so two remain, only the full inliner can inline this.

cameel replied:
Thanks. That's very useful to know. I'll update the docs here.

cameel replied:
Updated to `FullInliner`.

cameel committed:
Optimizer docs: Add missing info about steps that destroy the SSA and expression-split forms

cameel commented on docs/internals/optimizer.rst:
```diff
@@ -458,6 +462,14 @@ is transformed to
+
+To avoid unnecessary rewriting, it is recommended to run this step after StructuralSimplifier.
+
+May destroy the :ref:`expression-split <expression-splitter>` form.
```
I inferred the info about which steps preserve the expression-split and SSA forms from the current docs. Please double check that I did not make a mistake here. Since most steps seem to preserve the forms, I used the convention of stating it only if the step can destroy it.

For example this step (`ForLoopConditionIntoBody`) may destroy expression-split form because it would transform something like this:

```yul
for { ... } iszero(a) { ... } {
    ...
}
```

into this:
```yul
for { ... } 1 { ... } {
    if iszero(iszero(a)) { break }
}
```

`ExpressionSimplifier` also seems to mostly preserve it but also has some rules that don't. For example the `exp` rule will transform this:

```yul
let _0 := 0
let a := calldataload(_0)
let b := not(_0)
let c := exp(b, a)
sstore(a, c)
```

into this:
```yul
let _0 := 0
let a := calldataload(_0)
let b := 0xffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffff
let c := sub(iszero(and(a, 1)), and(a, 1))
sstore(a, c)
```

clonker replied:
I would phrase it a bit differently, what do you think about

```suggestion
The :ref:`expression-split <expression-splitter>` form is not guaranteed to be preserved under this step.
```

cameel commented on docs/internals/optimizer.rst:
```diff
@@ -1222,13 +1281,52 @@ code than more efficient code. In some cases, though, inlining a function
+    /* ... code of f, with replacements: a -> f_a, b -> f_b, c -> f_c */
+    let z := f_c
+
+During inlining, a heuristic is used to tell if the function call should be inlined or not.
+
+A function call will never be inlined if:
+
+- The function definition contains the ``leave`` opcode.
+- The function is large and inlining is not *aggressive* (details below).
+- The function definition contains any direct, recursive calls to itself or the call being inlined is recursive.
+  Note that indirect recursive calls (going through other functions) do not disqualify the function from inlining.
+- Argument expressions passed into the call can potentially have side effects.
+  This check is very broad, rejecting anything that is not simply a literal or an identifier, since
+  the condition is easy to satisfy by running ExpressionSplitter beforehand.
+
+Otherwise, if any of the following conditions is met, the call is inlined unconditionally:
+
+- The function is tiny.
+- The function is only ever called once.
+
+In the remaining cases the decision to inline or not depends on the size of the function.
+The size used here is not the actual bytecode size (which is not known to the optimizer), but rather
+a score calculated at the Yul level, which approximates that size.
+Inlining is performed if the size does not exceed a predefined threshold, which is a little higher
+if the inlining is aggressive and yet higher when the call has some constant arguments.
 
+The aggressive inlining is performed when:
+
+- The :ref:`yul-memoryguard` is present.
+- The function is not recursive (not even indirectly).
+- The legacy Yul->EVM transform is not used (i.e. the target is not the ``homestead`` :ref:`EVM version <evm-version>`).
```
This is a summary of what [`FullInliner::shallInline()`](https://github.com/ethereum/solidity/blob/v0.8.25/libyul/optimiser/FullInliner.cpp#L178-L238) does.

ekpyron replied:
I really don't think it is good to document things into this level of detail :-). There is little merit in it, it only leads to people *relying* on some behaviour we want to be able to change and it's a maintenance hassle.

cameel replied:
This is something I needed to know multiple times in the past and each time I just looked at the code of `shallInline()`, but it's nowhere as clear as having it spelled out in the docs and I never had time to dive into details so I actually never fully understood how it really worked. For example, that the `m_noInlineFunctions` check is simply a check for functions using `leave`. Or how some indirect recursive calls may actually be subject to inlining (there are three separate checks for recursive calls and it's not obvious how they differ). This time I needed it again to figure out why a function was not being inlined in one of my snippets. It always felt like something that should have been documented *somewhere*.

I could put it the docstring or inside `FullInliner::shallInline()`, but I really don't see much problem with having it in the docs proper. It's our main resource to understand how the optimizer works, not just for users but also for us and that fully justifies some extra effort spent keeping it up to date.

At the same time I don't think the extra detail really hurts users. It does not go into tiny details of how the inliner is implemented. It just describes the heuristic, which is useful to know when you're optimizing your code. But even then it omits all the specifics that could easily change, like numbers or details of the metric. There are step descriptions that are much more detailed than that. So if you're worried about users relying on them - we already have that problem. I don't think it's a good reason to avoid documenting things. I think that we in general lean too hard on that and I think it's one of the reasons why the code is so hard to get into.

In any case, I feel very strongly that the inliner heuristic should be documented.

ekpyron commented:
The main problem is: do you really want to guarantee that, assuming we satisfy these ordering requirements, arbitrary sequences don't lead to buggy cases :-)? We did miss one such case before and filed it as low severity bug, since custom sequences are rarely used and a "you-need-to-know-what-you-do" case reflected in lack of docs. If we document this in detail now, the situation changes - and we don't test any of this for correctness (and also can't - there's no way we can cover the space here with tests - we do fuzz this, but the fuzzer here can't catch everything...)

I'll in any case, still have a look and check if anything strikes me as odd since this affects https://github.com/ethereum/solidity/pull/15030.

We could also consider documenting this, but clearly stating that custom sequences are "at your own risk" anyways - even though that'd be a bit weird to start doing now out of the blue :-)... not sure yet :-).

ekpyron commented on docs/cheatsheet.rst:
```diff
@@ -168,3 +168,42 @@ Modifiers
+:ref:`u <unused-pruner>`                    Disambiguator
+:ref:`V <ssa-reverser>`                     Disambiguator                                                                                                         CommonSubexpressionEliminator, UnusedPruner
+:ref:`v <equivalent-function-combiner>`     Disambiguator, FunctionHoister                                                                                        UnusedPruner
+:ref:`x <expression-splitter>`                                                                    ForLoopConditionIntoBody
```
While it may be true, it's a stretch to assume that this *doesn't* depend on the `Disambiguator` - the idea is that *all* steps do and there's no value on trying to be more precise than that. The same goes for the block flattener and the function hoister and the for loop init rewriter  - basically the ``hgfo`` sequence we have due to that hardcoded and non-configurable in the sequence.
There's also not much point in omitting them for some steps as hard dependency, since disabling them is not a configuration option anyways :-).

ekpyron replied:
And: slightly overapproximating is fully harmless here - while understating may lead to damage :-).

cameel replied:
Note that the info about prerequisites is not something my PR introduces. It only fills in the missing bits from the docstrings and also organizes them in a table so that it's now easy to how spotty this is, but most of that info was already there. Anyone reading the docs now will already come away with an impression that some steps don't have this prerequisite, because otherwise, why would it be mentioned for some of them.

So I guess I should take it as a confirmation of my suggestion in https://github.com/ethereum/solidity/pull/15054#discussion_r1579863421 that it would be better to remove this info instead and add a general note that all steps should be assumed to depend on `hgfo`?

cameel replied:
I removed info about prerequisites from the optimizer docs and from the table. Now they're only in the docstrings. I also added a paragraph at the beginning of "Preprocessing" listing prerequisite steps and saying that all other steps implicitly depend on them.

cameel reviewed (commented):
> The main problem is: do you really want to guarantee that, assuming we satisfy these ordering requirements, arbitrary sequences don't lead to buggy cases :-)?

I really don't see how you can think that this PR changes anything in that regard. This long and detailed page with descriptions of optimizer steps already exists. We even have documented options for supplying arbitrary sequences, we already mostly document prerequisites, etc. The PR merely makes the info more consistent and summarizes it in a table. It does add some more detailed info from docstrings, but not that much of it and some steps already have even more detailed descriptions.

> We could also consider documenting this, but clearly stating that custom sequences are "at your own risk" anyways - even though that'd be a bit weird to start doing now out of the blue :-)... not sure yet :-).

Then we should just add such a warning. I think that the page is already more detailed than a typical user needs, but I don't think we should remove all the useful info from it. That would just hurt us. For me, for example, it's the primary resource I refer to when I need to understand how the optimizer works. Maybe it should not be a part of the main docs if that prevents us from documenting things, but having this info collected in one place like that is immensely useful. Thanks to this I know more about the optimizer than about some other components, e.g. the EVM transform. I wish we had even more pages about internals like that *somewhere*.

For the time being it's a part of the docs, so I think it should be kept up to date there until/if we decide to move it elsewhere.

cameel committed:
Optimizer docs: add missing prerequisites, missing bits of extra info from step docstrings and remove outdated remarks about steps being EVM-specific

cameel committed:
Optimizer docs: Remove detailed info about hard prerequisites for each step

- From user's perspective it's more useful to assume that all other steps implicitly depend on these prerequisites.
- For development the info is still available in docstrings of steps.

cameel commented on docs/internals/optimizer.rst:
```diff
@@ -853,19 +876,29 @@ It tries to match patterns like ``X + 0`` on each subexpression.
+
+This is mostly used so that other components do not have to rely on the DataflowAnalyzer.
+
+While the step destroys the :ref:`expression-split <expression-splitter>` form in the strict sense,
+it does not introduce any nested function calls, which may be good enough for many steps that require
+this property.
```
Also clarified this for `LiteralRematerialiser`, based on our discussion on the chat:

> You also don't need to consider the LiteralRematerializer to destroy expression-split form - it doesn't do so in a relevant way - even though it'd be interesting and to check if that makes any difference. I.e. if we find any case, in which this changes, and it's only due to the LiteralRematerializer, then it's likely ok (which we could then verify directly)

clonker replied:
Does any step actually require the ast to be in expression split form as prerequisite? Isn't it rather that some steps run potentially more efficiently if it is fulfilled? In that case I would phrase it more like

```suggestion
While the step destroys the :ref:`expression-split <expression-splitter>` form in the strict sense,
it does not introduce any nested function calls, which may still be good enough for many steps that require
this property to function efficiently.
```

clonker replied:
Impacts at least `EqualStoreEliminator`

clonker commented on docs/internals/optimizer.rst:
```diff
@@ -385,6 +385,20 @@ The preprocessing components perform transformations to get the program
+Some of the steps in this group should be considered prerequisites for all other components.
+Currently these are:
+
+- :ref:`disambiguator`
```
above we write `Some steps depend on properties ensured by ...`, which does not list disambiguator but the rest of the four. For consistency, I'd also add it there (line 349, for some reason, gh does not want me to add a review comment there)

clonker commented on docs/internals/optimizer.rst:
```diff
@@ -385,6 +385,20 @@ The preprocessing components perform transformations to get the program
+- :ref:`disambiguator`
+- :ref:`function-hoister`
+- :ref:`function-grouper`
+- :ref:`for-loop-init-rewriter`
```
Can we reorder the detailed description of the steps so that it reflects this order? Right now it is `Disambiguator`, `FunctionHoister`, `FunctionGrouper`, `ForLoopConditionIntoBody`, and only then `ForLoopInitRewriter`

clonker commented on docs/internals/optimizer.rst:
```diff
@@ -458,6 +472,12 @@ is transformed to
 This transformation can also be useful when paired with LoopInvariantCodeMotion, since
 invariants in the loop-invariant conditions can then be taken outside the loop.
 
+Loops that already have a literal constant as iteration condition are not transformed.
```
```suggestion
Loops that already have a literal constant or an individual identifier as iteration condition are not transformed.
```

clonker commented on docs/internals/optimizer.rst:
```diff
@@ -458,6 +472,12 @@ is transformed to
 
+Loops that already have a literal constant as iteration condition are not transformed.
+
+To avoid unnecessary rewriting, it is recommended to run this step after StructuralSimplifier.
```
How so, wouldn't the structural simplifier just remove for loops that have a `false` condition with its initialization? These cases would be ignored by the `ForLoopConditionIntoBody` step anyways

clonker commented on docs/internals/optimizer.rst:
```diff
@@ -832,10 +852,13 @@ value, one of them will always be unused. The UnusedPruner or the
-have a higher chance of expressions to be replaceable.
+The longer the values of variables are available, the higher the chance of an expression being replaceable.
+For this reason the step could potentially be more effective when used on code in pseudo-SSA form.
+This property, however, is not easily destroyed once established, and in most cases it is enough
```
It is not quite clear from the text what "this" property should be (pseudo ssa form vs long-livedness of variables' values), we could phrase it as 

```suggestion
The longevity of variable values' availability, however, is mostly preserved once established, and therefore in most cases it is enough
```

clonker commented on docs/internals/optimizer.rst:
```diff
@@ -832,10 +852,13 @@ value, one of them will always be unused. The UnusedPruner or the
+For this reason the step could potentially be more effective when used on code in pseudo-SSA form.
+This property, however, is not easily destroyed once established, and in most cases it is enough
+if SSATransform was used once at any point before running the step.
+The difference may also not be very significant in practice.
```
The difference of what vs what? If it is running SSA immediately beforehand vs just somewhere before, I suggest to drop this sentence

clonker commented on docs/internals/optimizer.rst:
```diff
@@ -853,19 +876,29 @@ It tries to match patterns like ``X + 0`` on each subexpression.
-in split or pseudo-SSA form.
+in :ref:`expression-split <expression-splitter>` or :ref:`pseudo-SSA <ssa-transform>` form.
+
+Some of the simplifications destroy the :ref:`expression-split <expression-splitter>` form.
```
As above I would phrase it as the expression split form not being guaranteed to be preserved

clonker commented on docs/internals/optimizer.rst:
```diff
@@ -875,9 +908,10 @@ LoadResolver
+is a constant ``<= 32``.
 
-Prerequisites: Disambiguator, ForLoopInitRewriter.
+Works best if the code is in SSA form.
```
but this would also invalidate the (strict) expression split form, I think

clonker replied:
impacts effectiveness of `EqualStoreEliminator`

clonker commented on docs/internals/optimizer.rst:
```diff
@@ -1018,6 +1047,8 @@ a structural level:
 
 This component uses the DataflowAnalyzer.
 
+LiteralRematerialiser should be run before this step.
```
it destroys the expression-split form e.g. if the condition of an `if`-statement is a function call but the body is empty.
`if (iszero(a)) {}` -> `pop(iszero(a))`

clonker commented on docs/internals/optimizer.rst:
```diff
@@ -1066,7 +1097,8 @@ declarations inside conditional branches will not be moved out of the loop.
 ExpressionSplitter and SSATransform should be run upfront to obtain better results.
 
-Prerequisites: Disambiguator, ForLoopInitRewriter, FunctionHoister.
+The transformation will not move loop-invariant condition out of the condition block of the loop.
```
```suggestion
The transformation will not move loop-invariant conditions out of the condition block of the loop.
```
